3p/iframe-messaging-client.js

```diff
@@ -88,37 +86,20 @@ export class IframeMessagingClient {
    * @private
    */
   setupEventListener_() {
-    listen(this.win_, 'message', message => {
+    listen(this.win_, 'message', event => {
       // Does it look a message from AMP?
-      if (message.source != this.hostWindow_) {
+      if (event.source != this.hostWindow_) {
         return;
       }
-      if (!message.data) {
-        return;
-      }
-      if (!startsWith(String(message.data), 'amp-')) {
+
+      const message = deserializeMessage(event.data);
+      if (!message || message.sentinel != this.sentinel_) {
         return;
       }
 
-      // See if we can parse the payload.
-      try {
-        const payload = JSON.parse(message.data.substring(4));
-        // Check the sentinel as well.
-        if (payload.sentinel == this.sentinel_ &&
-            this.callbackFor_[payload.type]) {
-          try {
-            // We should probably report exceptions within callback
-            const callback = this.callbackFor_[payload.type];
-            callback(payload);
-          } catch (err) {
-            user().error(
-                'IFRAME-MSG',
-                `- Error in registered callback ${payload.type}`,
-                err);
-          }
-        }
-      } catch (e) {
-        // JSON parsing failed. Ignore the message.
+      const callback = this.callbackFor_[message.type];
+      if (callback) {
+        callback(message);
```

Comment:
well, this callback function is supposed to be  provided by 3p. we should let them to handle.